maxq10xx/maxq10xx-wolfssl-client.c

```diff
@@ -0,0 +1,371 @@
+#include <wolfssl/wolfcrypt/rsa.h>
+#include <wolfssl/wolfcrypt/asn.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+
+#if (defined(WOLFSSL_MAXQ1065) || defined(WOLFSSL_MAXQ108X)) && \
+    defined(HAVE_PK_CALLBACKS) && defined(WOLF_CRYPTO_CB) && \
+    !defined(NO_PSK) && !defined(HAVE_EXTENDED_MASTER) && \
+    defined(NO_WOLFSSL_SERVER)
+
+/* -------------- */
+/* Configurations */
+/* -------------- */
+
+/* Please define this if you want wolfSSL's debug output */
+#define WANT_DEBUG
+
+/* Please set the server's address and the port it listens on */
+#define DEFAULT_SERVER "127.0.0.1"
+#define DEFAULT_PORT   11111
+
+/* ------------------------------------ */
+/* No modifications required below here */
+/* ------------------------------------ */
+
+/* identity is OpenSSL testing default for openssl s_client, keep same */
+static const char* kIdentityStr = "Client_identity";
+
+static WC_INLINE unsigned int my_psk_client_cb(WOLFSSL* ssl, const char* hint,
+        char* identity, unsigned int id_max_len, unsigned char* key,
+        unsigned int key_max_len)
+{
+    (void)ssl;
+    (void)hint;
+    (void)key;
+    (void)key_max_len;
+
+    /* see internal.h MAX_PSK_ID_LEN for PSK identity limit */
+    XSTRNCPY(identity, kIdentityStr, id_max_len);
+
+    /* For TLS 1.2, we indicate that MAXQ has the PSK. */
+    return USE_HW_PSK;
+}
+
+#ifdef WOLFSSL_TLS13
+static WC_INLINE unsigned int my_psk_client_cs_cb(WOLFSSL* ssl,
+        const char* hint, char* identity, unsigned int id_max_len,
+        unsigned char* key, unsigned int key_max_len, const char* ciphersuite)
+{
+    (void)ssl;
+    (void)hint;
+    (void)key;
+    (void)key_max_len;
+
+#ifdef WOLFSSL_PSK_MULTI_ID_PER_CS
+    /* Multiple calls for each cipher suite. First identity byte indicates the
+     * number of identites seen so far for cipher suite. */
+    if (identity[0] != 0) {
+        return 0;
+    }
+#endif
+
+    /* see internal.h MAX_PSK_ID_LEN for PSK identity limit */
+    XSTRNCPY(identity, kIdentityStr, id_max_len);
+    XSTRNCAT(identity, ciphersuite + XSTRLEN(ciphersuite) - 6, id_max_len);
+
+    /* For TLS 1.3, we just return an unmodified key. */
+    return 32;
+}
+#endif /* WOLFSSL_TLS13 */
+
+static void print_usage() {
+    fprintf(stderr, "usage: ./maxq10xx-wolfssl-client [-tls12|-tls13] "
+                    "[-ecc|-rsa|-psk] <cipher suite>\n");
+    fprintf(stderr, "Use `example/client/client -e` to get a list of cipher "
+                    "suites.\n");
+    fprintf(stderr, "Specifying a cipher suite is Optional.\n");
+}
+
+#define TLS12 1
+#define TLS13 2
+
+static int tls_version = 0;
+static int use_psk = 0;
+static char *key_file = NULL;
+static char *ciphersuite = NULL;
+
+static int cmd_line_parse(int argc, char** argv) {
+    const size_t tls_vers_len = 6;
+    const size_t alg_len = 4;
+
+    if ((argc != 3) && (argc != 4)) {
+        print_usage();
+        return -1;
+    }
+
+    if (strncmp("-tls12", argv[1], tls_vers_len) == 0) {
```

Comment:
Why have a fixed location for these command line arguments?

Reply:
Keeping the code as simple as possible outweighs having the flexibility of being able to have them anywhere.  MAXQ is meant for embedded systems so people reading this typically won't even have command line arguments in their code. Keeping it minimal seems appropriate.

Reply:
OK